### DISCUSSION TITLE
searching for control characters

### DISCUSSION THREAD

**Discussion by njamescouk** [BurntSushi/ripgrep#2578] [Q&A] [answered]
I haven't found a way to search for control characters using hex escapes,  specifically I'm looking for \x1f - unit separator - is this possible?

note I'm on windows 7 and doing rg <alt 31> works perfectly, however it fails, owing to windows problems, when called from a batch file. I'm thinking something like rg \x31 would work from a batch file.

**Comment by BurntSushi** [comment, marked as answer]
I don't know what `rg <alt 31>` means, but `rg '\x31'` should work:

```
$ printf "\x31" > haystack
$ xxd haystack
00000000: 31                                       1
$ rg '\x31' haystack
1:1
$ rg '\x31' haystack | xxd
00000000: 310a                                     1.
```

Note that this is on Linux. The important bit here is that ripgrep recognizes the `\x31` itself as an escape sequence and interprets it as matching the literal byte (in this case) `0x31`. You basically just need to be careful that your shell doesn't do something to mangle it. I'm not familiar with Windows shell quoting, so you'll need to figure that out.

**Reply by njamescouk** [reply]
edit: * this is incorrect *

the key combination alt 31 (I'm looking for unit separators) sends hex 1f to the editor, command line or whatever.

**Comment by njamescouk** [comment]
false alarm I'm afraid, looked at the files I was searching with a hex editor and the unit separators have been mangled to utf-8 on insertion :-(

**Reply by BurntSushi** [reply]
`\x31` is already valid UTF-8.
